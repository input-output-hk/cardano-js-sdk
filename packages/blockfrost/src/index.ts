export * from './blockfrostAssetProvider';
export * from './blockfrostTxSubmitProvider';
export * from './blockfrostNetworkInfoProvider';
export * from './blockfrostUtxoProvider';
export * from './blockfrostChainHistoryProvider';
export * from './blockfrostRewardsProvider';
export { Options } from '@blockfrost/blockfrost-js/lib/types';
export { BlockFrostAPI } from '@blockfrost/blockfrost-js';
