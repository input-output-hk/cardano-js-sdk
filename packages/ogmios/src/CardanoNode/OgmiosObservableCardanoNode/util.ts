import { Cardano, PointOrOrigin } from '@cardano-sdk/core';
import { Schema } from '@cardano-ogmios/client';

export const pointOrOriginToOgmios = (point: PointOrOrigin) =>
  point === 'origin'
    ? 'origin'
    : {
        hash: point.hash,
        slot: point.slot
      };

export const ogmiosToCorePoint = (point: Schema.Point) => ({
  hash: Cardano.BlockId(point.hash),
  slot: Cardano.Slot(point.slot)
});

export const ogmiosToCoreTip = (tip: Schema.Tip) => ({
  ...ogmiosToCorePoint(tip),
  blockNo: Cardano.BlockNo(tip.blockNo)
});

export const ogmiosToCoreTipOrOrigin = (tip: Schema.TipOrOrigin) => (tip === 'origin' ? tip : ogmiosToCoreTip(tip));

export const ogmiosToCorePointOrOrigin = (point: Schema.PointOrOrigin) =>
  point === 'origin' ? point : ogmiosToCorePoint(point);
