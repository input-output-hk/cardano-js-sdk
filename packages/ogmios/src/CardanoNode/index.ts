export * from './OgmiosCardanoNode';
export * from './OgmiosObservableCardanoNode';
