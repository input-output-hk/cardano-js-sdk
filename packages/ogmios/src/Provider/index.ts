export * from './TxSubmitProvider/index.js';
