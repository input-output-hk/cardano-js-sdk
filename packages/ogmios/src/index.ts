export * from './Provider';
export * from './CardanoNode';
export * from './util';
export * as ogmiosToCore from './ogmiosToCore';
export * as Ogmios from '@cardano-ogmios/client';
