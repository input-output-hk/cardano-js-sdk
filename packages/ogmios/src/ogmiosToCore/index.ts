export * from './block';
export * from './genesis';
export * from './types';
