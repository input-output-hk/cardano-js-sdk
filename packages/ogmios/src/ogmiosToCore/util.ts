import { BlockKind } from './types';
import { Schema } from '@cardano-ogmios/client';

export const BYRON_TX_FEE_COEFFICIENT = 43_946_000_000;
export const BYRON_TX_FEE_CONSTANT = 155_381_000_000_000;

export const isNativeScript = (script: Schema.Script): script is Schema.Native => 'native' in script;

export const isPlutusV1Script = (script: Schema.Script): script is Schema.PlutusV1 => 'plutus:v1' in script;

export const isPlutusV2Script = (script: Schema.Script): script is Schema.PlutusV2 => 'plutus:v2' in script;

export const isRequireAllOf = (nativeScript: Schema.ScriptNative): nativeScript is Schema.All =>
  typeof nativeScript === 'object' && 'all' in nativeScript;

export const isRequireAnyOf = (nativeScript: Schema.ScriptNative): nativeScript is Schema.Any =>
  typeof nativeScript === 'object' && 'any' in nativeScript;

export const isExpiresAt = (nativeScript: Schema.ScriptNative): nativeScript is Schema.ExpiresAt =>
  typeof nativeScript === 'object' && 'expiresAt' in nativeScript;

export const isStartsAt = (nativeScript: Schema.ScriptNative): nativeScript is Schema.StartsAt =>
  typeof nativeScript === 'object' && 'startsAt' in nativeScript;

export const isRequireNOf = (nativeScript: Schema.ScriptNative): nativeScript is Schema.NOf =>
  typeof nativeScript === 'object' && !Number.isNaN(Number(Object.keys(nativeScript)[0]));

export const isAlonzoOrAbove = (kind: BlockKind) => kind === 'babbage' || kind === 'alonzo';

export const isMaryOrAbove = (kind: BlockKind) => isAlonzoOrAbove(kind) || kind === 'mary';

export const isShelleyTx = (kind: BlockKind) => kind === 'shelley';
