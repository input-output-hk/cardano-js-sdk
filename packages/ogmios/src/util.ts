import { Cardano } from '@cardano-sdk/core';
import {
  ConnectionConfig,
  InteractionContext,
  InteractionType,
  ServerHealth,
  createInteractionContext
} from '@cardano-ogmios/client';
import { Logger } from 'ts-log';

/**
 * Converts an Ogmios connection URL to an Ogmios ConnectionConfig Object
 *
 * @param {URL} connectionURL Ogmios connection URL
 * @returns {ConnectionConfig} the ConnectionConfig Object
 */
export const urlToConnectionConfig = (connectionURL?: URL): ConnectionConfig => ({
  host: connectionURL?.hostname,
  port: connectionURL ? Number.parseInt(connectionURL.port) : undefined,
  tls: connectionURL?.protocol === 'wss'
});

type CreateInteractionContextOptions = {
  connection?: ConnectionConfig;
  interactionType?: InteractionType;
};

/**
 * Creates an Ogmios InteractionContext with close logic
 *
 * @param {Logger} logger Logger instance
 * @param {CreateInteractionContextOptions} options Options passed to the createInteractionContext function
 * @param {Function} onUnexpectedClose Optional callback to trigger an action when an unexpected close event occurs
 * @returns {Promise<InteractionContext>} Promise resolving an Ogmios InteractionContext
 */
export const createInteractionContextWithLogger = (
  logger: Logger,
  options?: CreateInteractionContextOptions,
  onUnexpectedClose?: () => Promise<void>
): Promise<InteractionContext> =>
  createInteractionContext(
    (error) => {
      logger.error({ error }, error.message);
    },
    async (code, reason) => {
      if (code === 1000) {
        logger.info({ code }, reason);
      } else {
        logger.error({ code }, 'Connection closed');
        await onUnexpectedClose?.();
      }
    },
    options
  );

export const ogmiosServerHealthToHealthCheckResponse = ({ lastKnownTip, networkSynchronization }: ServerHealth) => ({
  localNode: {
    ledgerTip: lastKnownTip,
    networkSync: Cardano.Percent(networkSynchronization)
  },
  ok: networkSynchronization > 0.99
});
