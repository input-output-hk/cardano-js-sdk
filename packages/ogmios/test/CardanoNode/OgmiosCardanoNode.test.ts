/* eslint-disable sonarjs/no-duplicate-string */
import { CardanoNodeErrors } from '@cardano-sdk/core';
import { Connection, createConnectionObject } from '@cardano-ogmios/client';
import { InvalidModuleState } from '@cardano-sdk/util';
import { OgmiosCardanoNode } from '../../src';
import { createMockOgmiosServer, listenPromise, serverClosePromise } from '../mocks/mockOgmiosServer';
import { getRandomPort } from 'get-port-please';
import { dummyLogger as logger } from 'ts-log';
import http from 'http';
describe('OgmiosCardanoNode', () => {
  let mockServer: http.Server;
  let connection: Connection;
  let node: OgmiosCardanoNode;
  const moduleName = 'OgmiosCardanoNode';

  beforeAll(async () => {
    connection = createConnectionObject({ port: await getRandomPort() });
  });
  describe('not initialized and started', () => {
    beforeAll(async () => {
      mockServer = createMockOgmiosServer({
        stateQuery: { eraSummaries: { response: { success: true } }, systemStart: { response: { success: true } } }
      });
      await listenPromise(mockServer, connection.port);
      node = new OgmiosCardanoNode(connection, logger);
    });
    afterAll(async () => {
      await serverClosePromise(mockServer);
    });

    it('eraSummaries rejects with not initialized error', async () => {
      await expect(node.eraSummaries()).rejects.toThrowError(
        new CardanoNodeErrors.NotInitializedError('eraSummaries', moduleName)
      );
    });
    it('systemStart rejects with not initialized error', async () => {
      await expect(node.systemStart()).rejects.toThrowError(
        new CardanoNodeErrors.NotInitializedError('systemStart', moduleName)
      );
    });
    it('stakeDistribution rejects with not initialized error', async () => {
      await expect(node.stakeDistribution()).rejects.toThrowError(
        new CardanoNodeErrors.NotInitializedError('stakeDistribution', moduleName)
      );
    });
    it('shutdown rejects with not initialized error', async () => {
      await expect(node.shutdown()).rejects.toThrowError(InvalidModuleState);
    });
  });
  describe('initialized and started', () => {
    describe('eraSummaries', () => {
      describe('success', () => {
        beforeAll(async () => {
          mockServer = createMockOgmiosServer({
            stateQuery: { eraSummaries: { response: { success: true } }, systemStart: { response: { success: true } } }
          });
          await listenPromise(mockServer, connection.port);
          node = new OgmiosCardanoNode(connection, logger);
          await node.initialize();
          await node.start();
        });
        afterAll(async () => {
          await node.shutdown();
          await serverClosePromise(mockServer);
        });
        it('resolves if successful', async () => {
          const res = await node.eraSummaries();
          expect(res).toMatchSnapshot();
        });
      });
      describe('failure', () => {
        beforeAll(async () => {
          mockServer = createMockOgmiosServer({
            stateQuery: {
              eraSummaries: { response: { failWith: { type: 'unknownResultError' }, success: false } },
              systemStart: { response: { success: true } }
            }
          });
          await listenPromise(mockServer, connection.port);
          node = new OgmiosCardanoNode(connection, logger);
          await node.initialize();
          await node.start();
        });
        afterAll(async () => {
          await node.shutdown();
          await serverClosePromise(mockServer);
        });

        it('rejects with errors thrown by the service', async () => {
          await expect(node.eraSummaries()).rejects.toThrowError(
            CardanoNodeErrors.CardanoClientErrors.UnknownResultError
          );
        });
      });
    });
    describe('systemStart', () => {
      describe('success', () => {
        beforeAll(async () => {
          mockServer = createMockOgmiosServer({ stateQuery: { systemStart: { response: { success: true } } } });
          await listenPromise(mockServer, connection.port);
          node = new OgmiosCardanoNode(connection, logger);
          await node.initialize();
          await node.start();
        });
        afterAll(async () => {
          await node.shutdown();
          await serverClosePromise(mockServer);
        });
        it('resolves if successful', async () => {
          const res = await node.systemStart();
          expect(res).toMatchSnapshot();
        });
      });
      describe('failure', () => {
        beforeAll(async () => {
          mockServer = createMockOgmiosServer({
            stateQuery: { systemStart: { response: { failWith: { type: 'queryUnavailableInEra' }, success: false } } }
          });
          await listenPromise(mockServer, connection.port);
          node = new OgmiosCardanoNode(connection, logger);
          await node.initialize();
          await node.start();
        });
        afterAll(async () => {
          await node.shutdown();
          await serverClosePromise(mockServer);
        });

        it('rejects with errors thrown by the service', async () => {
          await expect(node.systemStart()).rejects.toThrowError(
            CardanoNodeErrors.CardanoClientErrors.QueryUnavailableInCurrentEraError
          );
        });
      });
    });
    describe('healthCheck', () => {
      describe('success', () => {
        beforeAll(async () => {
          mockServer = createMockOgmiosServer({
            healthCheck: { response: { success: true } }
          });
          await listenPromise(mockServer, connection.port);
          node = new OgmiosCardanoNode(connection, logger);
          await node.initialize();
          await node.start();
        });
        afterAll(async () => {
          await node.shutdown();
          await serverClosePromise(mockServer);
        });
        it('returns ok if successful', async () => {
          const res = await node.healthCheck();
          expect(res.ok).toBe(true);
        });
      });
      describe('failure', () => {
        beforeAll(async () => {
          mockServer = createMockOgmiosServer({
            healthCheck: {
              response: {
                failWith: new Error('Some error'),
                success: false
              },
              skipInvocations: 1
            }
          });
          await listenPromise(mockServer, connection.port);
          node = new OgmiosCardanoNode(connection, logger);
          await node.initialize();
          await node.start();
        });
        afterAll(async () => {
          await node.shutdown();
          await serverClosePromise(mockServer);
        });

        it('returns not ok if the Ogmios server throws an error', async () => {
          const res = await node.healthCheck();
          expect(res.ok).toBe(false);
        });
      });
    });
    describe('stakeDistribution', () => {
      describe('success', () => {
        beforeAll(async () => {
          mockServer = createMockOgmiosServer({
            stateQuery: { stakeDistribution: { response: { success: true } } }
          });
          await listenPromise(mockServer, connection.port);
          node = new OgmiosCardanoNode(connection, logger);
          await node.initialize();
          await node.start();
        });
        afterAll(async () => {
          await node.shutdown();
          await serverClosePromise(mockServer);
        });
        it('resolves if successful', async () => {
          const res = await node.stakeDistribution();
          expect(res).toMatchSnapshot();
        });
      });
      describe('failure', () => {
        beforeAll(async () => {
          mockServer = createMockOgmiosServer({
            stateQuery: {
              stakeDistribution: { response: { failWith: { type: 'queryUnavailableInEra' }, success: false } },
              systemStart: { response: { success: true } }
            }
          });
          await listenPromise(mockServer, connection.port);
          node = new OgmiosCardanoNode(connection, logger);
          await node.initialize();
          await node.start();
        });
        afterAll(async () => {
          await node.shutdown();
          await serverClosePromise(mockServer);
        });

        it('rejects with errors thrown by the service', async () => {
          await expect(node.stakeDistribution()).rejects.toThrowError(
            CardanoNodeErrors.CardanoClientErrors.QueryUnavailableInCurrentEraError
          );
        });
      });
    });
    describe('shutdown', () => {
      beforeAll(async () => {
        mockServer = createMockOgmiosServer({ stateQuery: { systemStart: { response: { success: true } } } });
        await listenPromise(mockServer, connection.port);
      });
      afterAll(async () => {
        await serverClosePromise(mockServer);
      });
      beforeEach(async () => {
        node = new OgmiosCardanoNode(connection, logger);
        await node.initialize();
        await node.start();
      });
      it('shuts down successfully', async () => {
        await expect(node.shutdown()).resolves.not.toThrow();
      });

      it('throws when querying after shutting down', async () => {
        await node.shutdown();
        await expect(node.systemStart()).rejects.toThrow(CardanoNodeErrors.NotInitializedError);
      });
    });
  });
});
