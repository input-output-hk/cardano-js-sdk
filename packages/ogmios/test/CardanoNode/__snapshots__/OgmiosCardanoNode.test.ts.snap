// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OgmiosCardanoNode initialized eraSummaries success resolves if successful 1`] = `
Array [
  Object {
    "parameters": Object {
      "epochLength": 21600,
      "slotLength": 20000,
    },
    "start": Object {
      "slot": 0,
      "time": 2017-09-23T21:44:51.000Z,
    },
  },
  Object {
    "parameters": Object {
      "epochLength": 432000,
      "slotLength": 1000,
    },
    "start": Object {
      "slot": 1598400,
      "time": 2018-09-28T21:44:51.000Z,
    },
  },
]
`;

exports[`OgmiosCardanoNode initialized stakeDistribution success resolves if successful 1`] = `
Map {
  "pool1la4ghj4w4f8p4yk4qmx0qvqmzv6592ee9rs0vgla5w6lc2nc8w5" => Object {
    "stake": Object {
      "pool": 10098109508n,
      "supply": 40453712883332027n,
    },
    "vrf": "4e4a2e82dc455449bf5f1f6d249470963cf97389b5dc4d2118fe21625f50f518",
  },
  "pool1lad5j5kawu60qljfqh02vnazxrahtaaj6cpaz4xeluw5xf023cg" => Object {
    "stake": Object {
      "pool": 14255969766n,
      "supply": 40453712883332027n,
    },
    "vrf": "474a6d2a44b51add62d8f2fd8fe80abc722bf84478479b617ad05b39aaa84971",
  },
  "pool1llugtz5r4t6m7xz6es4qu7cszllm5y3uvx3ast5a9jzlv7h3xdu" => Object {
    "stake": Object {
      "pool": 98763124501826n,
      "supply": 40453712883332027n,
    },
    "vrf": "dc1c0fd7d2fd95b6e9bf0e50ab5cb722edbd7d6e85b7d53323884d429ec6a83c",
  },
  "pool1lu6ll4rcxm92059ggy6uym2p804s5hcwqyyn5vyqhy35kuxtn2f" => Object {
    "stake": Object {
      "pool": 1494933206n,
      "supply": 40453712883332027n,
    },
    "vrf": "4a13d5e99a1868788057bf401fdb4379b7846290dd948918839981088059a564",
  },
}
`;

exports[`OgmiosCardanoNode initialized systemStart success resolves if successful 1`] = `2017-09-23T21:44:51.000Z`;
